README.md
scala-leaderboard
=================

Leaderboards backed by Redis in Scala, http://redis.io.

Builds off ideas proposed in http://blog.agoragames.com/2011/01/01/creating-high-score-tables-leaderboards-using-redis/.

Installation
============

Make sure your redis server is running! Redis configuration is outside the scope of this README, but 
check out the Redis documentation, http://redis.io/documentation.

scala-leaderboard uses Scala Build Tool, https://github.com/harrah/xsbt/wiki, for building and testing. Clone the source from GitHub and build using:

    dczarnecki-agora:scala-leaderboard dczarnecki$ sbt package
    [info] Set current project to default (in build file:/Users/dczarnecki/projects/scala-leaderboard/)
    [info] Packaging /Users/dczarnecki/projects/scala-leaderboard/target/scala-2.9.0.final/scala-leaderboard_2.9.0-2.0.0.jar ...
    [info] Done packaging.
    [success] Total time: 1 s, completed Aug 5, 2011 9:20:44 AM  

Testing
=======

You can run the tests using:

    sbt test

Compatibility
============

The project has been developed using Scala version 2.9.0.1 and Scala Build Tool version 0.10

Usage
============

Create a new leaderboard or attach to an existing leaderboard named 'highscores':

    scala> var highscore_lb = new Leaderboard("highscores", "localhost", 6379, 25)
    leaderboard: com.agoragames.leaderboard.Leaderboard = com.agoragames.leaderboard.Leaderboard@2f8d604f
    
Add members to your leaderboard using rankMember:

    scala> for (i <- 1 to 10) {
         | highscore_lb.rankMember("member_" + i, i)
         | }
    
You can call rankMember with the same member and the leaderboard will be updated automatically.

Get some information about your leaderboard:

    scala> highscore_lb.totalMembers
    res4: Option[Int] = Some(10)
    
    scala> highscore_lb.totalPages
    res5: Int = 1
    
Get some information about a specific member(s) in the leaderboard:

    scala> highscore_lb.scoreFor("member_4")
    res6: Option[Double] = Some(4.0)

    scala> highscore_lb.rankFor("member_4")
    res7: Option[Int] = Some(7)

    scala> highscore_lb.rankFor("member_10")
    res8: Option[Int] = Some(1)
    
Get page 1 in the leaderboard:

    scala> highscore_lb.leaders(1)
    res9: java.util.List[(String, Double, Int)] = [(member_10,10.0,1), (member_9,9.0,2), (member_8,8.0,3), (member_7,7.0,4), (member_6,6.0,5), (member_5,5.0,6), (member_4,4.0,7), (member_3,3.0,8), (member_2,2.0,9), (member_1,1.0,10)]
    
Add more members to your leaderboard:

    scala> for (i <- 50 to 95) {
         | highscore_lb.rankMember("member_" + i, i)
         | }

    scala> highscore_lb.totalPages
    res11: Int = 3
    
Get an "Around Me" leaderboard for a member:

    scala> highscore_lb.aroundMe("member_53")
    res13: java.util.List[(String, Double, Int)] = [(member_65,65.0,31), (member_64,64.0,32), (member_63,63.0,33), (member_62,62.0,34), (member_61,61.0,35), (member_60,60.0,36), (member_59,59.0,37), (member_58,58.0,38), (member_57,57.0,39), (member_56,56.0,40), (member_55,55.0,41), (member_54,54.0,42), (member_53,53.0,43), (member_52,52.0,44), (member_51,51.0,45), (member_50,50.0,46), (member_10,10.0,47), (member_9,9.0,48), (member_8,8.0,49), (member_7,7.0,50), (member_6,6.0,51), (member_5,5.0,52), (member_4,4.0,53), (member_3,3.0,54), (member_2,2.0,55)]
    
Get rank and score for an arbitrary list of members (e.g. friends):

    scala> highscore_lb.rankedInList(Array("member_1", "member_62", "member_67"))
    res19: java.util.List[(String, Double, Int)] = [(member_1,1.0,56), (member_62,62.0,34), (member_67,67.0,29)]
    
Other useful methods:

    deleteLeaderboard: Delete the current leaderboard  
    removeMember(member): Remove a member from the leaderboard
    totalMembers: Total # of members in the leaderboard
    totalPages: Total # of pages in the leaderboard given the leaderboard's page_size	
    changeScoreFor(member, score): Change the score for a member by some amount delta (delta could be positive or negative)
    rankFor(member): Retrieve the rank for a given member in the leaderboard
    scoreFor(member): Retrieve the score for a given member in the leaderboard
    checkMember(member): Check to see whether member is in the leaderboard
    scoreAndRankFor(member): Retrieve the score and rank for a member in a single call
    removeMembersInScoreRange(minScore, maxScore): Remove members from the leaderboard within a score range

Performance Metrics
===================

10 million sequential scores insert:

    scala> benchmark(1) {
         | report("sequential insert") {
         | for (i <- 1 to 10000000) {
         | highscore_lb.rankMember("member_" + i, i)
         | }
         | } ::
         | Nil
         | }
    
         Name Time(s)
         ===============================
         sequential insert    963.581000
         res4: List[(String, Long)] = List((sequential insert,963581))

Average time to request an arbitrary page from the leaderboard:
    
    scala> benchmark(50000) {
         | report("leaderboard page requests") {
         | highscore_lb.leaders(scala.util.Random.nextInt(highscore_lb.totalPages) + 1)
         | } ::
         | Nil
         | }
    Name Time(s)
    ===============================
    leaderboard page requests 241.492000
    res19: List[(String, Long)] = List((leaderboard page requests,241492))

    scala> println(241.492000 / 50000)
    0.00482984
        
Future Ideas
============
  
Contributing to scala-leaderboard
=================================
 
* Check out the latest master to make sure the feature hasn't been implemented or the bug hasn't been fixed yet
* Check out the issue tracker to make sure someone already hasn't requested it and/or contributed it
* Fork the project
* Start a feature/bugfix branch
* Commit and push until you are happy with your contribution
* Make sure to add tests for it. This is important so I don't break it in a future version unintentionally.
* Please try not to mess with the build files, version, or history. If you want to have your own version, or is otherwise necessary, that is fine, but please isolate to its own commit so I can cherry-pick around it.

Copyright
============

Copyright (c) 2011 David Czarnecki. See LICENSE.txt for further details.



CHANGELOG.md
# leaderboard 2.0.0 (in progress)

  * Change `addMember` to `rankMember`
  * Added `deleteLeaderboard` and `deleteLeaderboardNamed`
  * Added transaction support for `scoreAndRankFor`, `leaders`, `aroundMe` and `rankedInList`



src/test/scala/com/agoragames/LeaderboardSpec.scala
package com.agoragames

import org.scalatest.Spec
import org.scalatest.BeforeAndAfterEach
import org.scalatest.BeforeAndAfterAll
import org.scalatest.matchers.ShouldMatchers
import org.scalatest.junit.JUnitRunner
import org.junit.runner.RunWith

import com.redis._

@RunWith(classOf[JUnitRunner])
class LeaderboardSpec extends Spec 
                        with ShouldMatchers
                        with BeforeAndAfterEach
                        with BeforeAndAfterAll {

	val redisClient = new RedisClient("localhost", 6379)
    var leaderboard = new Leaderboard("leaderboard_name", "localhost", 6379, 25)

	override def beforeEach = {
		redisClient.flushdb
	}

	override def afterEach = {
	}

	override def afterAll = {
		redisClient.disconnect
		leaderboard.disconnect
	}
	
	private def addMembersToLeaderboard(totalMembers: Int) = {
		for (i <- 1 to totalMembers) {
			leaderboard.addMember("member_" + i, i)
		}
	}
	
    describe("version") {
      it("should be the correct version") {
          
          leaderboard.version should equal("1.0.0")
      }
    }
    
    describe("constructor") {
        it("should have the correct parameters") {
            leaderboard.leaderboardName should equal("leaderboard_name")
            leaderboard.pageSize should equal(25)
        }
    }
        
    describe("totalMembers and totalMembersIn") {
        it("should return the correct number of members for totalMembers") {
            leaderboard.totalMembers should equal(Some(0))
            leaderboard.totalMembers.get should equal(0)
        }
        
        it("should return the correct number of members for totalMembersIn") {
            leaderboard.totalMembersIn("leaderboard_name") should equal(Some(0))
            leaderboard.totalMembersIn("leaderboard_name").get should equal(0)
        }
    }
    
    describe("addMember and addMemberTo") {
        it("should be able to add a member to the leaderboard using addMember") {        
            leaderboard.addMember("member", 1337) should equal(true)
            leaderboard.totalMembersIn("leaderboard_name").get should equal(1)

            leaderboard.addMember("member", 1338) should equal(false)
            leaderboard.totalMembersIn("leaderboard_name").get should equal(1)
        }

        it("should be able to add a member to the leaderboard using addMemberTo") {
            leaderboard.addMemberTo("leaderboard_name", "member", 1337) should equal(true)
            leaderboard.totalMembersIn("leaderboard_name").get should equal(1)

            leaderboard.addMemberTo("leaderboard_name", "member", 1338) should equal(false)
            leaderboard.totalMembersIn("leaderboard_name").get should equal(1)

            leaderboard.addMemberTo("leaderboard_name", "another_member", 1339) should equal(true)
            leaderboard.totalMembersIn("leaderboard_name").get should equal(2)
         }
    }
    
    describe("totalPages and totalPagesIn") {
        it("should return the correct number of pages in the leaderboard using totalPages for a single page") {
            addMembersToLeaderboard(5)
            
            leaderboard.totalPages("leaderboard_name") should equal(1)
        }

        it("should return the correct number of pages in the leaderboard using totalPages") {
            addMembersToLeaderboard(LeaderboardDefaults.DEFAULT_PAGE_SIZE + 2)
            
            leaderboard.totalPages("leaderboard_name") should equal(2)
        }

        it("should return the correct number of pages in the leaderboard using totalPagesIn") {
            addMembersToLeaderboard(LeaderboardDefaults.DEFAULT_PAGE_SIZE + 2)
            
            leaderboard.totalPagesIn("leaderboard_name", LeaderboardDefaults.DEFAULT_PAGE_SIZE) should equal(2)
        }
    }
    
    // RedisClient does not currently support zcount.
    // describe("totalMembersInScoreRange and totalMembersInScoreRangeIn") {
    //     it("should return correct number of members for totalMembersInScoreRange") {
    //         addMembersToLeaderboard(5)
    //         
    //         leaderboard.totalMembersInScoreRange(2, 4) should equal(3)
    //     }
    // }
    
    describe("scoreFor and scoreforIn") {
        it("should return the correct score for a member using scoreFor") {
            addMembersToLeaderboard(5)
            
            leaderboard.scoreFor("member_3").get should equal(3.0)
        }

        it("should return the correct score for a member using scoreForIn") {
            addMembersToLeaderboard(5)
            
            leaderboard.scoreForIn("leaderboard_name", "member_3").get should equal(3.0)
        }
    }

    describe("changeScoreFor and changeScoreforIn") {
        it("should return the correct score for a member using changeScoreFor") {
            addMembersToLeaderboard(5)
            
            leaderboard.changeScoreFor("member_3", 6).get should equal(9.0)
            leaderboard.changeScoreFor("member_3", -3).get should equal(6.0)
        }

        it("should return the correct score for a member using changeScoreForIn") {
            addMembersToLeaderboard(5)
            
            leaderboard.changeScoreForIn("leaderboard_name", "member_3", 6).get should equal(9.0)
            leaderboard.changeScoreForIn("leaderboard_name", "member_3", -3).get should equal(6.0)
        }
    }
    
    describe("checkMember and checkMemberIn") {
        it("should return whether or not a member is in the leaderboard using checkMember") {
            addMembersToLeaderboard(5)

            leaderboard.totalMembers.get should equal(5)
            leaderboard.checkMember("member_3") should equal(true)
            leaderboard.checkMember("member_10") should equal(false)
        }

        it("should return whether or not a member is in the leaderboard using checkMemberIn") {
            addMembersToLeaderboard(5)
            
            leaderboard.checkMemberIn("leaderboard_name", "member_3") should equal(true)
            leaderboard.checkMemberIn("leaderboard_name", "member_10") should equal(false)
        }
    }
    
    describe("rankFor and rankForIn") {
        it("should return the correct rank for rankFor") {
            addMembersToLeaderboard(5)
            
            leaderboard.rankFor("member_4").get should equal(2)
            leaderboard.rankFor("member_4", true).get should equal(1)
        }
    }
    
    describe("scoreAndRankFor and scoreAndRankForIn") {
        it("should return the correct rank and score for scoreAndRankFor") {
            addMembersToLeaderboard(5)
            
            val dataMap: scala.collection.mutable.HashMap[String, Object] = leaderboard.scoreAndRankFor("member_1")
            
            dataMap("member") should equal("member_1")
            dataMap("score").asInstanceOf[Option[Double]].get should equal(1.0)
            dataMap("rank").asInstanceOf[Option[Int]].get should equal(5)
        }
    }
    
    describe("removeMembersInScoreRange and removeMembersInScoreRangeIn") {
        it("should return the correct number of members removed for removeMembersInScoreRange") {
            addMembersToLeaderboard(5)

            leaderboard.totalMembers.get should equal(5)

            leaderboard.addMember("cheater_1", 100)
            leaderboard.addMember("cheater_2", 101)
            leaderboard.addMember("cheater_3", 102)

            leaderboard.totalMembers.get should equal(8)

            leaderboard.removeMembersInScoreRange(100, 102)

            leaderboard.totalMembers.get should equal(5)            
        }
    }
}

src/test/scala/com/agoragames/leaderboard/LeaderboardDataSpec.scala
package com.agoragames.leaderboard

import org.scalatest.Spec
import org.scalatest.BeforeAndAfterEach
import org.scalatest.BeforeAndAfterAll
import org.scalatest.matchers.ShouldMatchers
import org.scalatest.junit.JUnitRunner
import org.junit.runner.RunWith

@RunWith(classOf[JUnitRunner])
class LeaderboardDataSpec extends Spec 
                        with ShouldMatchers
                        with BeforeAndAfterEach
                        with BeforeAndAfterAll {
                            
    describe("constructor") {
        it("should have the correct parameters") {
            var leaderboardData = new LeaderboardData("member", 5.0, 5)
            
            leaderboardData.member should equal("member")
            leaderboardData.score should equal(5.0)
            leaderboardData.rank should equal(5)
        }
    }
}

src/main/scala/com/agoragames/leaderboard/Leaderboard.scala
package com.agoragames.leaderboard

import com.redis._

object LeaderboardDefaults {
    val VERSION = "2.0.0"
    val DEFAULT_PAGE_SIZE = 25
    val DEFAULT_REDIS_HOST = "localhost"
    val DEFAULT_REDIS_PORT = 6379   
}

class Leaderboard(leaderboardNameParam: String, 
        host: String = LeaderboardDefaults.DEFAULT_REDIS_HOST, 
        port: Int = LeaderboardDefaults.DEFAULT_REDIS_PORT, 
        pageSizeParam: Int = LeaderboardDefaults.DEFAULT_PAGE_SIZE,
        redisOptions: scala.collection.mutable.HashMap[String, Object] = scala.collection.mutable.HashMap("host" -> LeaderboardDefaults.DEFAULT_REDIS_HOST, "port" -> LeaderboardDefaults.DEFAULT_REDIS_PORT.asInstanceOf[AnyRef])) {
    private var redisClient: RedisClient = _;
    
    if (redisOptions.get("redis_connection") != None) {
        redisClient = redisOptions.get("redis_connection").get.asInstanceOf[RedisClient]
    } else {
        redisClient = new RedisClient(host, port)
    }
    
    val version = LeaderboardDefaults.VERSION
    val leaderboardName: String = leaderboardNameParam
    
    var pageSize: Int = pageSizeParam
    
    if (pageSize < 1) {
        pageSize = LeaderboardDefaults.DEFAULT_PAGE_SIZE
    }
    
    def disconnect: Boolean = {
        redisClient.disconnect
    }
    
    def deleteLeaderboard: Option[Int] = {
        this.deleteLeaderboardNamed(this.leaderboardName)    
    }
    
    def deleteLeaderboardNamed(leaderboardName: String): Option[Int] = {
        redisClient.del(leaderboardName)
    }
        
    def totalMembers: Option[Int] = {
        this.totalMembersIn(this.leaderboardName)
    }
        
    def totalMembersIn(leaderboardName: String): Option[Int] = {
        redisClient.zcard(leaderboardName)
    }

    def rankMember(memberName: String, score: Double): Boolean = {
        this.rankMemberIn(this.leaderboardName, memberName, score)
    }
    
    def rankMemberIn(leaderboardName: String, memberName: String, score: Double): Boolean = {
        redisClient.zadd(leaderboardName, score, memberName)
    }
    
    def removeMember(memberName: String): Boolean = {
        this.removeMemberFrom(this.leaderboardName, memberName)
    }
    
    def removeMemberFrom(leaderboardName: String, memberName: String): Boolean = {
        redisClient.zrem(leaderboardName, memberName)
    }
    
    def totalPages: Int = {
        this.totalPagesIn(this.leaderboardName, this.pageSize)
    }
    
    def totalPagesIn(leaderboardName: String, pageSize: Int): Int = {
        scala.math.ceil(this.totalMembersIn(leaderboardName).get.asInstanceOf[Float] / pageSize.asInstanceOf[Float]).asInstanceOf[Int]
    }

    // RedisClient does not currently support zcount.
    // def totalMembersInScoreRange(minScore: Double, maxScore: Double): Int = {
    //     this.totalMembersInScoreRangeIn(this.leaderboardName, minScore, maxScore)
    // }
    // 
    // def totalMembersInScoreRangeIn(leaderboardName: String, minScore: Double, maxScore: Double): Int = {
    //     redisClient.zcount(leaderboardName, minScore, maxScore)
    // }
    
    def scoreFor(member: String): Option[Double] = {
        this.scoreForIn(this.leaderboardName, member)
    }
    
    def scoreForIn(leaderboardName: String, member: String): Option[Double] = {
        redisClient.zscore(leaderboardName, member)
    }
    
    def changeScoreFor(member: String, score: Double): Option[Double] = {
        this.changeScoreForIn(this.leaderboardName, member, score)
    }
    
    def changeScoreForIn(leaderboardName: String, member: String, score: Double): Option[Double] = {
        redisClient.zincrby(leaderboardName, score, member)
    }

    def checkMember(member: String): Boolean = {
        this.checkMemberIn(this.leaderboardName, member)
    }

    def checkMemberIn(leaderboardName: String, member: String): Boolean = {
        !(redisClient.zscore(leaderboardName, member) == None)
    }

    def rankFor(member: String, useZeroIndexForRank: Boolean = false): Option[Int] = {
        this.rankForIn(this.leaderboardName, member, useZeroIndexForRank)
    }
    
    def rankForIn(leaderboardName: String, member: String, useZeroIndexForRank: Boolean = false): Option[Int] = {
        if (useZeroIndexForRank) {
            redisClient.zrank(leaderboardName, member, true)            
        } else {
            // This feels "not elegant"
            Some(new java.lang.Integer(redisClient.zrank(leaderboardName, member, true).get + 1))
        }
    }
    
    def scoreAndRankFor(member: String, useZeroIndexForRank: Boolean = false): scala.collection.mutable.HashMap[String, Object] = {
        this.scoreAndRankForIn(this.leaderboardName, member, useZeroIndexForRank)
    }
    
    def scoreAndRankForIn(leaderboardName: String, member: String, useZeroIndexForRank: Boolean = false): scala.collection.mutable.HashMap[String, Object] = {
        var dataMap = scala.collection.mutable.HashMap.empty[String, Object]
        var responses: List[Any] = redisClient.pipeline { transaction =>
            transaction.zscore(leaderboardName, member)
            transaction.zrank(leaderboardName, member, true)
        }.get
                
        dataMap += ("member" -> member)
        dataMap += ("score" -> responses(0).asInstanceOf[Option[Double]])
        if (!useZeroIndexForRank) {
            dataMap += ("rank" -> Some(responses(1).asInstanceOf[Option[Int]].get + 1))
        } else {
            dataMap += ("rank" -> Some(responses(1).asInstanceOf[Option[Int]]))            
        }     
                
        dataMap
    }

    def removeMembersInScoreRange(minScore: Double, maxScore: Double): Option[Int] = {
        this.removeMembersInScoreRangeIn(this.leaderboardName, minScore, maxScore)
    }
    
    def removeMembersInScoreRangeIn(leaderboardName: String, minScore: Double, maxScore: Double): Option[Int] = {
        redisClient.zremrangebyscore(leaderboardName, minScore, maxScore)
    }
    
    def leaders(currentPage: Int, withScores: Boolean = true, withRank: Boolean = true, useZeroIndexForRank: Boolean = false, pageSize: Int = LeaderboardDefaults.DEFAULT_PAGE_SIZE) = {
        this.leadersIn(this.leaderboardName, currentPage, withScores, withRank, useZeroIndexForRank, pageSize)
    }
    
    def leadersIn(leaderboardName: String, currentPageParam: Int, withScores: Boolean = true, withRank: Boolean = true, useZeroIndexForRank: Boolean = false, pageSize: Int = LeaderboardDefaults.DEFAULT_PAGE_SIZE): java.util.List[(String, Double, Int)] = {
        var currentPage: Int = currentPageParam
        
        if (currentPage < 1) {
            currentPage = 1
        }
        
        if (currentPage > totalPagesIn(leaderboardName, pageSize)) {
            currentPage = totalPagesIn(leaderboardName, pageSize)
        }
        
        var indexForRedis: Int = currentPage - 1
        var startingOffset: Int = (indexForRedis * pageSize)
        if (startingOffset < 0) {
            startingOffset = 0
        }
        var endingOffset: Int = (startingOffset + pageSize) - 1
                
        var rawLeaderData = redisClient.zrange(leaderboardName, startingOffset, endingOffset, RedisClient.DESC)
        var massagedLeaderData: java.util.List[(String, Double, Int)] = new java.util.ArrayList[(String, Double, Int)]
        
        if (rawLeaderData != None) {
            var responses = redisClient.pipeline { transaction =>
                for (leader <- rawLeaderData.get) {
                    transaction.zscore(leaderboardName, leader)
                    transaction.zrank(leaderboardName, leader, true)
                }
            }.get
        
            for (leaderIndex <- rawLeaderData.get.indices) {
                var rank = responses(leaderIndex * 2 + 1).asInstanceOf[Some[Int]].get
            
                if (!useZeroIndexForRank) {
                    rank += 1
                }
            
                massagedLeaderData.add((rawLeaderData.get(leaderIndex), responses(leaderIndex * 2).asInstanceOf[Some[Double]].get, rank))
            }
        }
        
        massagedLeaderData        
    }
    
    def aroundMe(member: String, withScores: Boolean = true, withRank: Boolean = true, useZeroIndexForRank: Boolean = false, pageSize: Int = LeaderboardDefaults.DEFAULT_PAGE_SIZE): java.util.List[(String, Double, Int)] = {
        this.aroundMeIn(this.leaderboardName, member, withScores, withRank, useZeroIndexForRank, pageSize)      
    }
    
    def aroundMeIn(leaderboardName: String, member: String, withScores: Boolean = true, withRank: Boolean = true, useZeroIndexForRank: Boolean = false, pageSize: Int = LeaderboardDefaults.DEFAULT_PAGE_SIZE): java.util.List[(String, Double, Int)] = {
        var reverseRankForMember: Int = redisClient.zrank(leaderboardName, member, true).get

        var startingOffset: Int = reverseRankForMember - (pageSize / 2)
        if (startingOffset < 0) {
          startingOffset = 0
        }
        var endingOffset = (startingOffset + pageSize) - 1

        var rawLeaderData = redisClient.zrange(leaderboardName, startingOffset, endingOffset, RedisClient.DESC)
        var massagedLeaderData: java.util.List[(String, Double, Int)] = new java.util.ArrayList[(String, Double, Int)]        

        if (rawLeaderData != None) {
            var responses = redisClient.pipeline { transaction =>
                for (leader <- rawLeaderData.get) {
                    transaction.zscore(leaderboardName, leader)
                    transaction.zrank(leaderboardName, leader, true)
                }
            }.get
        
            for (leaderIndex <- rawLeaderData.get.indices) {
                var rank = responses(leaderIndex * 2 + 1).asInstanceOf[Some[Int]].get
            
                if (!useZeroIndexForRank) {
                    rank += 1
                }
            
                massagedLeaderData.add((rawLeaderData.get(leaderIndex), responses(leaderIndex * 2).asInstanceOf[Some[Double]].get, rank))
            }
        }

        massagedLeaderData
    }
    
    def rankedInList(members: Array[String], withScores: Boolean = true, useZeroIndexForRank: Boolean = false): java.util.List[(String, Double, Int)] = {
        rankedInListIn(this.leaderboardName, members, withScores, useZeroIndexForRank)
    }
    
    def rankedInListIn(leaderboardName: String, members: Array[String], withScores: Boolean = true, useZeroIndexForRank: Boolean = false): java.util.List[(String, Double, Int)] = {
        var ranksForMembers: java.util.List[(String, Double, Int)] = new java.util.ArrayList[(String, Double, Int)]
        
        var responses = redisClient.pipeline { transaction =>
            for (member <- members) {
                transaction.zscore(leaderboardName, member)
                transaction.zrank(leaderboardName, member, true)
            }
        }.get
    
        for (memberIndex <- members.indices) {
            var rank = responses(memberIndex * 2 + 1).asInstanceOf[Some[Int]].get
        
            if (!useZeroIndexForRank) {
                rank += 1
            }
        
            ranksForMembers.add((members(memberIndex), responses(memberIndex * 2).asInstanceOf[Some[Double]].get, rank))
        }
        
        
        ranksForMembers
    }
}